python/pyspark/ml/util.py

```diff
@@ -61,32 +66,82 @@ def _randomUID(cls):
 
 
 @inherit_doc
-class MLWriter(object):
+class BaseReadWrite(object):
+    """
+    Base class for MLWriter and MLReader. Stores information about the SparkContext
+    and SparkSession.
+
+    .. versionadded:: 2.3.0
+    """
+
+    def __init__(self):
+        self.sparkSession = None
+
+    def context(self, sqlContext):
+        """
+        Sets the SQL context to use for saving.
+
+        .. note:: Deprecated in 2.1 and will be removed in 3.0, use session instead.
+        """
+        # raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
+        self.sparkSession = sqlContext.sparkSession
+        return self
+
+    def session(self, sparkSession):
+        """Sets the Spark Session to use for saving."""
+        # raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
+        self.sparkSession = sparkSession
+        return self
+
+    def getOrCreateSparkSession(self):
+        if self.sparkSession is None:
+            self.sparkSession = SparkSession.builder.getOrCreate()
+        return self.sparkSession
+
+    @property
+    def sc(self):
+        return self.getOrCreateSparkSession().sparkContext
+
+
+@inherit_doc
+class MLWriter(BaseReadWrite):
     """
     Utility class that can save ML instances.
 
     .. versionadded:: 2.0.0
     """
 
+    def __init__(self):
+        super(MLWriter, self).__init__()
+        self.shouldOverwrite = False
+
     def save(self, path):
         """Save the ML instance to the input path."""
+        if isinstance(self, JavaMLWriter):
+            self.saveImpl(path)
+        else:
+            if self.shouldOverwrite:
+                os.remove(path)
+            self.saveImpl(path)
+
+    def saveImpl(self, path):
         raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
 
     def overwrite(self):
         """Overwrites if the output path already exists."""
-        raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
+        self.shouldOverwrite = True
```

Comment:
return self (to allow builder pattern); always match Scala API where possible & where it makes sense